lib/backend/watchersyncer/watchersyncer.go

```diff
@@ -0,0 +1,200 @@
+	Process(*model.KVPair) ([]*model.KVPair, error)
+
+	// SyncStarting is called when syncer is starting a full sync for the associated resource
+	// type.  That means it is first going to list current resources and then watch for any updates.
+	// If the processor maintains a private internal cache, then the cache should be cleared at
+	// this point since the cache will be re-populated from the sync.
+	SyncStarting()
+}
+
+// New creates a new multiple Watcher-backed api.Syncer.
+func New(client api.Client, resourceTypes []ResourceType, callbacks api.SyncerCallbacks) api.Syncer {
+	rs := &watcherSyncer{
+		watcherCaches: make([]*watcherCache, len(resourceTypes)),
+		results:       make(chan interface{}, 2000),
+		callbacks:     callbacks,
+	}
+	for i, r := range resourceTypes {
+		rs.watcherCaches[i] = newWatcherCache(client, r, rs.results)
+	}
+	return rs
+}
+
+// watcherSyncer implements the api.Syncer interface.
+type watcherSyncer struct {
+	status        api.SyncStatus
+	watcherCaches []*watcherCache
+	results       chan interface{}
+	numSynced     int
+	callbacks     api.SyncerCallbacks
+}
+
+func (rs *watcherSyncer) Start() {
+	log.Info("Start called")
+	go rs.run()
+}
+
+// Send a status update and store the status.
+func (rs *watcherSyncer) sendStatusUpdate(status api.SyncStatus) {
+	log.WithField("Status", status).Info("Sending status update")
+	rs.callbacks.OnStatusUpdated(status)
+	rs.status = status
+}
+
+// run implements the main syncer loop that loops forever receiving watch events and translating
+// to syncer updates.
+func (rs *watcherSyncer) run() {
+	log.Info("Running - sending status update")
+	rs.sendStatusUpdate(api.WaitForDatastore)
+	log.Info("Starting watcher caches")
+	for _, wc := range rs.watcherCaches {
+		go wc.run()
+	}
+
+	log.Info("Starting main watch processing loop")
+	var updates []api.Update
+	for {
+		// Block until there is data.
+		result := <-rs.results
+
+		// Process the data - this will append the data in subsequent calls, and action
+		// it if we hit a non-update event.
+		updates := rs.processResult(updates, result)
+
+		// Append results into the one update until we either flush the channel or we
+		// hit our fixed limit per update.
+	consolidatationloop:
+		for ii := 0; ii < maxUpdatesToConsolidate; ii++ {
+			select {
+			case next := <-rs.results:
+				updates = rs.processResult(updates, next)
+			default:
+				break consolidatationloop
+			}
+		}
+
+		// Perform final processing (pass in a nil result) before we loop and hit the blocking
+		// call again.
+		updates = rs.processResult(updates, nil)
+	}
+
+}
+
+// Process a result from the result channel.  We don't immediately action updates, but
+// instead start grouping them together so that we can send a larger single update to
+// Felix.
+func (rs *watcherSyncer) processResult(updates []api.Update, result interface{}) []api.Update {
+
+	// Final call of the interation, send any updates if we've collated some.
```

Comment:
Would it be cleaner to split this out into its own function? e.g. `sendUpdates`?

Seems like we only call this with `nil` once so not sure why we need to cram it into this function.

Reply:
aha!   good spot!  I think I'd restructured the code and that was a side effect.

Reply:
Done